File: zivid_samples/scripts/sample_capture_and_save_frame.py
#!/usr/bin/env python

import rospy
import rosnode
import dynamic_reconfigure.client
from zivid_camera.srv import *
from std_msgs.msg import String


class Sample:
    def __init__(self):
        rospy.init_node("sample_capture_and_save_frame_py", anonymous=True)

        rospy.loginfo("Starting sample_capture_and_save_frame.py")

        rospy.wait_for_service("/zivid_camera/capture_and_save_frame", 30.0)

        self.capture_and_save_frame_service = rospy.ServiceProxy(
            "/zivid_camera/capture_and_save_frame", CaptureAndSaveFrame
        )

        self.enable_diagnostic()
        self.enable_first_acquistion()

    def capture(self):
        rospy.loginfo("Calling capture_and_save_frame service")
        file_path = "/tmp/capture_py.zdf"
        self.capture_and_save_frame_service(file_path)

    def enable_diagnostic(self):
        rospy.loginfo("Enabling the diagnostics mode")
        settings_client = dynamic_reconfigure.client.Client("/zivid_camera/settings/")
        settings_config = {"diagnostics_enabled": True}
        settings_client.update_configuration(settings_config)

    def enable_first_acquistion(self):
        rospy.loginfo("Enabling the first acquisition")
        acquisition_0_client = dynamic_reconfigure.client.Client(
            "/zivid_camera/settings/acquisition_0"
        )
        acquisition_0_config = {"enabled": True}
        acquisition_0_client.update_configuration(acquisition_0_config)


if __name__ == "__main__":
    s = Sample()
    s.capture()
    rospy.spin()

File: zivid_samples/src/sample_capture_and_save_frame.cpp
#include <zivid_camera/SettingsAcquisitionConfig.h>
#include <zivid_camera/SettingsConfig.h>
#include <zivid_camera/CaptureAndSaveFrame.h>
#include <dynamic_reconfigure/Reconfigure.h>
#include <dynamic_reconfigure/client.h>
#include <ros/ros.h>

#define CHECK(cmd)                                                                                                     \
  do                                                                                                                   \
  {                                                                                                                    \
    if (!cmd)                                                                                                          \
    {                                                                                                                  \
      throw std::runtime_error{ "\"" #cmd "\" failed!" };                                                              \
    }                                                                                                                  \
  } while (false)

namespace
{
const ros::Duration default_wait_duration{ 30 };

void capture_and_save_frame()
{
  ROS_INFO("Calling capture_and_save_frame service");
  zivid_camera::CaptureAndSaveFrame capture_and_save_frame;
  std::string file_path = "/tmp/capture_cpp.zdf";
  capture_and_save_frame.request.file_path = file_path;
  CHECK(ros::service::call("/zivid_camera/capture_and_save_frame", capture_and_save_frame));
}

void enable_diagnostics()
{
  dynamic_reconfigure::Client<zivid_camera::SettingsConfig> settings_client("/zivid_camera/"
                                                                            "settings/");

  // To initialize the settings_config object we need to load the default configuration from the server.
  // The default values of settings depends on which Zivid camera model is connected.
  zivid_camera::SettingsConfig settings_config;
  CHECK(settings_client.getDefaultConfiguration(settings_config, default_wait_duration));

  ROS_INFO("Enabling the diagnostics mode");
  settings_config.diagnostics_enabled = true;
  CHECK(settings_client.setConfiguration(settings_config));
}

void enable_first_acquisition()
{
  dynamic_reconfigure::Client<zivid_camera::SettingsAcquisitionConfig> acquisition_0_client("/zivid_camera/settings/"
                                                                                            "acquisition_0/");

  // To initialize the acquisition_0_config object we need to load the default configuration from the server.
  // The default values of settings depends on which Zivid camera model is connected.
  zivid_camera::SettingsAcquisitionConfig acquisition_0_config;
  CHECK(acquisition_0_client.getDefaultConfiguration(acquisition_0_config, default_wait_duration));

  ROS_INFO("Enabling the first acquisition");
  acquisition_0_config.enabled = true;
  CHECK(acquisition_0_client.setConfiguration(acquisition_0_config));
}

}  // namespace

int main(int argc, char** argv)
{
  ros::init(argc, argv, "sample_capture_and_save_frame_cpp");
  ros::NodeHandle n;

  ROS_INFO("Starting sample_capture_and_save_frame.cpp");

  CHECK(ros::service::waitForService("/zivid_camera/capture_and_save_frame", default_wait_duration));

  ros::AsyncSpinner spinner(1);
  spinner.start();

  enable_diagnostics();

  enable_first_acquisition();

  ROS_INFO("Calling capture_and_save_frame");

  capture_and_save_frame();

  ros::waitForShutdown();

  return 0;
}